matrix-dendrite: init at 0.3.11

mjlbach opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Dendrite is coming closer to a stable release. The package seems to build fine, and I was able to setup a home server.

I also made a basic monolithic configuration with the default settings autogenerated from the dendrite yaml in accordance with the latest settings RFC. (Thanks @Infinisil for your help!)

Some thoughts/goals

- [ ] Should we match the build flags [dendrite uses](https://github.com/matrix-org/dendrite/blob/master/build.sh#L22)
- [x] figure out what to do with all of the generically named bins (I don't think this matters too much, since most people will likely use the service anyways)
- [x] NixOS Module

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

mjlbach committed:
matrix-dendrite: init at 0.3.11

mjlbach committed:
maintainers: add mjlbach to the matrix team

mjlbach committed:
matrix-dendrite: add nixos module

mjlbach commented:
Work on the module is here: https://github.com/mjlbach/nixpkgs/tree/init_matrix_dendrite_module

SuperSandro2000 commented:
This is a semi-automatic executed nixpkgs-review which does not build all packages (e.g. lumo, tensorflow or pytorch)
If you find some bugs or got suggestions for further things to search or run please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 109561` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and are new build failure:</summary>
  <ul>
<li>matrix-dendrite: log <a href="https://termbin.com/uue5">https://termbin.com/uue5</a></li>
  </ul>
</details>

mjlbach commented:
Hmm, let me look into the build failure on darwin. That shouldn't be happening.

SuperSandro2000 commented:
> Hmm, I can't reproduce the failing test on my machine. Looking into it.

Keep in mind that I have sandbox enabled. Probably related to that.

Ma27 commented:
> Keep in mind that I have sandbox enabled. Probably related to that.

Is darwin even supported by upstream?

> I also made a basic monolithic configuration with the default settings autogenerated from the dendrite yaml in accordance with the latest settings RFC

There are two things I'm missing though (though feel free to correct me if I'm wrong :)):

* As shown in an [example in rfc42](https://github.com/NixOS/rfcs/blob/master/rfcs/0042-config-option.md#detailed-design), settings that are not expected to be used in 99% of the deployments should be set with a `mkDefault`.
* The RFC doesn't prohibit explicitly defined options in the `settings`-submodule: in some cases I'd prefer to have explicit options set, *at least* for e.g. network ports.
* I'm not sure if it's a good idea to just use all defaults from the upstream config, especially since it seems (to me at least) to be a reference to all possible configuration options:
  * By default, no kafka is used AFAIU, however it's configured in your defaults.
  * I'm not really sure if we want to hard-code any third-party servers as trusted (as option default it may be fine though since it can be seen in the manual then).
  * This may also apply to some other sections that I've missed during a brief review.
* Have you checked whether it would be sufficient to use DynamicUser in conjunction with StateDirectory from systemd rather than configuring a new user? This is not a requirement, though I'd be curious if that would suffice :)

mjlbach commented:
> Is darwin even supported by upstream?

Yes! It's [mentioned](https://github.com/matrix-org/dendrite/blob/master/docs/INSTALL.md#install-kafka-polylith-only) the installation instructions at least. FWIW it seems to work fine locally (for me). 

> By default, no kafka is used AFAIU, however it's configured in your defaults.

 I'm using [Naffka](https://github.com/NixOS/nixpkgs/blob/9fbe3acdf03f7d6e074e5f08f54816a1b5b2b162/nixos/modules/services/misc/matrix-dendrite.nix#L126). It's just configured under the kafka section.

> As shown in an example in rfc42, settings that are not expected to be used in 99% of the deployments should be set with a mkDefault.
>The RFC doesn't prohibit explicitly defined options in the settings-submodule: in some cases I'd prefer to have explicit options set, at least for e.g. network ports.
> I'm not sure if it's a good idea to just use all defaults from the upstream config, especially since it seems (to me at least) to be a reference to all possible configuration options: 
> I'm not really sure if we want to hard-code any third-party servers as trusted (as option default it may be fine though since it can be seen in the manual then).

If I delete the non-pertinent sections of the config from settings, can people still configure these options? (This is my first nixos module). 

Which upstream defaults should we keep then? And which one's should we make options? Right now I'm assuming port, and servers should be options, are there any others?

> Have you checked whether it would be sufficient to use DynamicUser in conjunction with StateDirectory from systemd rather than configuring a new user? This is not a requirement, though I'd be curious if that would suffice :)

I can try! Do you have a reference I can base this off of?

rnhmjoj commented:
> The RFC doesn't prohibit explicitly defined options in the settings-submodule: in some cases I'd prefer to have explicit options set, at least for e.g. network ports.

On the contrary, I find modules where all the configuration is dumped into a `settings` option pretty sad. I think there should be NixOS options for the most frequently used and mandatory settings. The type system and assertions are extremely useful for catching common errors, plus the can also provide a bit of documentation with examples.

mjlbach commented:
> On the contrary, I find modules where all the configuration is dumped into a settings option pretty sad. 

Sorry to make you sad 😆 I will do the work to fix this, this was just my first pass!

rnhmjoj commented:
> Sorry to make you sad 😆 I will do the work to fix this, this was just my first pass!

Of course, sorry: I didn't mean to offend your work.

mjlbach commented:
No offense taken, I was just prefacing this with some context that I might need more feedback this first time. I read through the source where they read settings, and tested it. This seems to be the minimal configuration necessary. Should I replace settings with this, and make options for server_name and trusted_third_party_id_servers?

```
version: 1
global:
  server_name: localhost
  trusted_third_party_id_servers:
  - matrix.org
  - vector.im
  kafka:
    use_naffka: true
client_api:
  external_api:
    listen: http://[::]:8071
federation_api:
  external_api:
    listen: http://[::]:8072
media_api:
  external_api:
    listen: http://[::]:8074
sync_api:
  external_api:
    listen: http://[::]:8073
logging:
- type: file
  level: info
  params:
    path: /home/michael/.cache/log/dendrite
```

I can add additional args for the ports in the systemd service, and also add these as options (so we'd have a total of 5, including dataDir).  These aren't read from the config but from the config file.

```
httpBindAddr   = flag.String("http-bind-address", ":8008", "The HTTP listening port for the server")
httpsBindAddr  = flag.String("https-bind-address", ":8448", "The HTTPS listening port for the server")
```

The other options I would probably want to include are 3 for paths to keys (for a total of 8), right now you can just copy the keys over to the directory made for the matrix-dendrite user.

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,342 @@
+        '';
+      };
+      settings = lib.mkOption {
+        type = lib.types.submodule {
+          freeformType = settingsFormat.type;
```
Not that it matters very much, but I'm seeing this a lot. Freeform modules aren't needed if the submodule doesn't declare any options. In this case this is equivalent to just
```nix
{
  type = settingsFormat.type;
}
```

mjlbach replied:
Out of curiosity, when is it better to declare options under submodules vs options under the toplevel  `options.services.matrix-dendrite` ? I want to understand this better, and hopefully contribute a bit to some tutorials and documentation.

mjlbach replied:
And upon re-reading everything and realizing that options is a required attribute of types.submodule, everything works! Thanks!

Infinisil replied:
For `settings`, a submodule is used to delimit where values can be freely set using `freeformType`. Within a `freeformType` submodule, you can declare values without having associated options. Whereas outside the submodule, every value needs to have an associated option (unless the parent module also has a `freeformType` set).

mjlbach replied:
Thank you! That makes sense. One other question: Settings itself is ultimately written (in it's entirety) as the configuration yaml (dendrite.yaml). There are probably other settings that I would want to declare without associated options which would be used to populate the command line flags to the daemon, which I'm currently doing with options (since if I put these under settings, they will be incorrectly written to the yaml). Is there a convention for multiple submodules that you had in mind, when you have multiple different configuration files for example?

mjlbach commented:
@Ma27 @rnhmjoj I believe I addressed most of the concerns, with the exception of switching to DynamicUser iw/ systemd StateDirectory. Do these default options look ok? I tried to expose the most common settings as options, and the config itself is fairly minimal now.

rnhmjoj commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,125 @@
+        RemainAfterExit = true;
+      };
+      script = lib.strings.concatStringsSep " " [
+        "generate-keys"
```
I guess this generates a self-signed cert for the homeserver, don't they [have to be signed](https://github.com/matrix-org/synapse/blob/master/docs/MSC1711_certificates_FAQ.md) by a CA to federate?
If so, there's should probably be an option to provide one, or maybe use ACME.

mjlbach replied:
Yes, that's a good point. I'll look into using ACME.

mjlbach replied:
I see acme requires a reverse proxy set up. Synapse doesn't include this in the service, but just documents how to set everything up. Should I configure an option for setting up acme here or document it?

mjlbach replied:
I just grep'd, it looks like all (but web servers ofc) rely on documenting how to use acme. Are we still using docgen, or can I write the documentation in markdown somehow?

rnhmjoj replied:
You're right, with matrix-synapse I'll simply disable tls and let the reverse proxy do it: this is what the [docs](https://github.com/matrix-org/synapse/blob/master/docs/reverse_proxy.md) recommends.
A simple solution could be to add options to specify the paths of the tls keypair and generate the self-signed cert only when tls is enabled but the options are unset. (I don't think there are existing settings in dendrite.yaml because these are passed from the command line)

Instead, the `matrix_key.pem` key is always required. It should only be generated if `global.private_key` or `global.old_private_keys` are unset (see [docs] https://github.com/matrix-org/dendrite/blob/master/docs/serverkeyformat.md).

mjlbach replied:
Should we not enable https by default and gate that behind specifying the https port? Or should I make enabling https a separate option?

mjlbach replied:
Also, given that old_private_keys can be converted to private_keys, is it worth adding that option? I know the RFC stressed not adding too many.

rnhmjoj replied:
> Should we not enable https by default and gate that behind specifying the https port? Or should I make enabling https a separate option?

Oh right, there's also this: It's not clear to me what these options are doing. Are both http and https enabled by default? How do you disable one listener?

> Also, given that old_private_keys can be converted to private_keys, is it worth adding that option? I know the RFC stressed not adding too many.

Uhm, if you like I think it could be done without defining a new option: 
1. make `private_key` a `types.nullOr types.path` and set it to `null` by default. 
2. set `settings.old_private_keys = lib.mkDefault null` in the `config` section
3. put a gate like this
```
lib.optionalString  (cfg.settings.old_private_keys == null && cfg.settings.private_key == null)
  "--private-key matrix_key.pem"
```

Anyway, this is certainly subjective, but I don't think ~8 options are too many: compare this to the 66 options of matrix-synapse module...

pacien replied:
Note that Dendrite's readme file also __recommends__ using a reverse proxy like nginx and letting it handle TLS, just like for Synapse: https://github.com/matrix-org/dendrite#requirements

mjlbach replied:
@rnhmjoj I addressed the http/https on a local branch that I have yet to push. Now I'm leaning towards remove the https option due to the following:

@pacien I personally will also be using it with a reverse proxy. That being said, maybe we just remove the https option and the user can enable it manually via cfg.settings if they want? There's not a precedent for setting up nginx as part of a module, so I assume it's ok to just document how to do it?

mjlbach replied:
So, this is getting a little messy. I just pushed but the derivation currently doesn't work.

These lines: https://github.com/mjlbach/nixpkgs/blob/9b3162789114b2e22fe804932dd37d8e5a01d596/nixos/modules/services/misc/matrix-dendrite.nix#L94-L98

cause errors

```
while evaluating the attribute 'values' at /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:538:9:
while evaluating the attribute 'values' at /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:637:7:
while evaluating anonymous function at /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:524:28, called from /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:524:17:
while evaluating definitions from `/home/michael/Repositories/nix/nixpkgs/nixos/modules/services/misc/matrix-dendrite.nix':
while evaluating 'dischargeProperties' at /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:596:25, called from /home/michael/Repositories/nix/nixpkgs/lib/modules.nix:525:137:
while evaluating the attribute 'value' at /home/michael/Repositories/nix/nixpkgs/lib/types.nix:386:60:
value is a string with context while a list was expected, at /home/michael/Repositories/nix/nixpkgs/nixos/modules/services/misc/matrix-dendrite.nix:89:23
```

Also, how do we create namespaces with the new settings RFC?

mjlbach replied:
Nvm fixed

mjlbach commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,125 @@
+      '';
+    };
+    httpPort = lib.mkOption {
+      type = lib.types.int;
```
Done.

mjlbach commented:
Note to self, also should make disallowing registration an option (probably default?)

mjlbach commented:
Ok, I think this is *mostly* good now. Some thoughts:

* The key generation service adds a lot of complexity (mkMerge and the conditions), and probably won't/shouldn't be used by most people because:
  - You should be using a reverse proxy anyways instead of the built-in TLS
  - You should be generating/storing your private key outside of nix (and if in the nix store, via an external tool like agenix)

That makes me tempted to just remove that systemd service

* I removed old_private_keys because it adds complexity, this is a new matrix server and it's not that hard to convert your old keys and path them in via path.

* Right now, you can enable/disable allowing users to sign up via settings.client_api.registration_disabled. I didn't make  registration_shared_secret an option (which allows *you* to log in when registration_disabled = true, because it would share the registration secret in plain text in the store. This can always be added in via settings if someone wants it.

Any other thoughts?

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+          ] ++ lib.optionals (cfg.tls_key != null) [
+            "--tls-key ${cfg.tls_key}"
+          ]);
+          ExecReload = "${pkgs.util-linux}/bin/kill -HUP $MAINPID";
```
`kill` should be in `PATH`, so you can omit the absolute path of it

szlend replied:
This doesn't work for me.

```
matrix-dendrite.service: Failed to locate executable kill: No such file or directory
matrix-dendrite.service: Failed at step EXEC spawning kill: No such file or directory
```

mjlbach replied:
@szlend You mean as it does now? Or omitting the absolute path?

szlend replied:
Yeah, I'm saying currently (without the full path) reload doesn't work for me.

mjlbach replied:
So `${pkgs.util-linux}/bin/kill` doesn't work, but does `kill` by itself?

mjlbach replied:
(sorry ninja edit)

szlend replied:
I didn't try to fix it, but `kill` by itself is definitely not enough in this instance.

mjlbach replied:
Changed it to use pkgs.coreutils like other modules

Infinisil replied:
Good to know! Will remember this :)

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+            RemainAfterExit = true;
+          };
+          script = lib.strings.concatStringsSep " " ([
+            "generate-keys"
```
I suggest doing this the same as the other service, so removing the `path = [ matrix-dendrite ]` and using
```nix
{
  serviceConfig.ExecStart = [ "${matrix-dendrite}/bin/generate-keys" ... ];
}
```
instead

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+    (lib.mkIf
+      (cfg.enable &&
+        cfg.generate_private_key == true || cfg.tls_cert != null ||
+        cfg.tls_key != null)
```
What exactly is the logic behind this condition? And shouldn't it be `cfg.tls_cert == null || cfg.tls_key == null` instead?

mjlbach replied:
The reason was that cfg.tls_cert is both the path to the key you want to generate, and the path used to use provided keys. I'm not super happy with how it currently is, and am leaning a bit towards just removing the key generation service.

Infinisil replied:
I guess you meant to say
> The reason was that cfg.tls_cert is both the path to the key you want to generate or the one provided by the user

But I'm still very confused about this explanation. How about adding an explanation for how it works to the `tlsCert`, `tlsKey` and/or `generatePrivateKey` options?

I think such a service is valuable if it reduces the need for manual interaction.

mjlbach replied:
The generatePrivateKey option reduces the need for manual interaction upon first time setup, I don't think most people will use tlsCert/tslKey because if you do a reverse proxy, this will be handled by nginx.

Would it be excessive to add ad generateTLS option (or generateTLSKey and generateTLSCert options) to make this more clear?

mjlbach replied:
I think the changes make it clearer. Let me know what you think.

Infinisil replied:
Hmm, doesn't the TLS key and cert have some relation between them? Like, the cert needs to be derived from the key, so it doesn't make any sense to generate a key but have a fixed cert not depending on the key? And is the private key in any way related?

Also, in the current module you'll get a `null` error when setting `generateTls{Key,Cert} = true` but not setting `tls{Key,Cert}` to non-null.

mjlbach replied:
Ah, you're right. I simplified the code a bit thank to this, now you have to pass in both. I think you can use your own private_key not related to TLS.

Is there a way to make the option generateTls (the new one I just pushed) depend on tlsKey and tlsCert being non-null? I think the way i have it now, it will silently fail (but not error).

Infinisil replied:
How about
```nix
{
  config = {
    services.matrix-dendrite.tlsCert = mkIf cfg.generateTls "/var/lib/matrix-dendrite/server.cert";
    services.matrix-dendrite.tlsKey = mkIf cfg.generateTls "/var/lib/matrix-dendrite/server.key";
  };
}
```

mjlbach replied:
Great idea! Added

mjlbach commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+      default = "matrix-dendrite";
+      description = ''
+        The directory where matrix-dendrite stores its stateful data such as
+        certificates, media and uploads.
```
Went with the former option, unless you think the other way is better!

mjlbach commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+            default = [ "matrix.org" "vector.im" ];
+            description = ''
+              Lists of domains that the server will trust as identity
+              servers to verify third
```
Late night :P

mjlbach commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+      default = { };
+      description = ''
+        Configuration for dendrite, see:
+        <link xlink:href="https://github.com/matrix-org/dendrite"/>
```
I put a direct link to the example yaml, which is the only documentation other than the source that I could find.

mjlbach commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+        after = [
+          "network.target"
+        ] ++ lib.optionals (cfg.generate_private_key == true) [
+          "matrix-dendrite-key-generator.service"
```
Oh, that's good to know! Fixed.

mjlbach commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+          server_name = lib.mkOption {
+            type = lib.types.str;
+            example = "example.com";
+            default = config.networking.hostName;
```
Hmm, should I just leave the default blank then?

Pacman99 replied:
maybe config.networking.domain? I don't know if that's always set though.

Infinisil replied:
Is this a required option? If so, we can leave it blank, which then forces the user to set it

mjlbach replied:
Yes, sounds good! Wasn't sure if I had to do anything special to have a nice error message.

mjlbach replied:
Fixed

mjlbach commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,179 @@
+  };
+
+  config = lib.mkMerge [
+    (lib.mkIf cfg.enable {
```
This is so much cleaner! Thanks :)

mjlbach commented:
I ended up setting settings.client_api.registration_disabled to default true, which means you *won't* be able to log into your own server by default without either `services.matrix-dendrite.settings.client_api.registration_disabled = true;` or `services.matrix-dendrite.settings.client_api.registration_shared_secret = "mysecretpassword;"

This is more secure, maybe a bit annoying though. I also don't like that it's store-readable, so I didn't make it an option.

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,171 @@
+            default = true;
+            description = ''
+              Whether to disable user registration to the server
+              without the shared secret.
```
If setting such a shared secret is a relatively common operation, I suggest adding another option (outside of `settings`) for securely specifying `client_api.registration_shared_secret` with a file. E.g.:
```nix
{
  options.services.matrix-dendrite.registrationSecretFile = mkOption {
    type = types.nullOr types.path;
  };

  config = mkIf (cfg.registrationSecretFile != null) {
    services.matrix-dendrite.settings.client_api.registration_shared_secret = "NIXOS_REGISTRATION_SECRET";
 
    systemd.services.matrix-dendrite.script = ''
      registrationSecret=$(cat ${cfg.registrationSecretFile})
      configFile=$(mktemp)
      sed "s/NIXOS_REGISTRATION_SECRET/$registrationSecret/" ${configurationYaml} > "$configFile"
      dendrite-monolith-server --config "$configFile"
    '';
  };
}
```

mjlbach replied:
I made the changes, but right now it will fail if cfg.registrationSecretFile is null

mjlbach replied:
Think I fixed it, pushing soon

Infinisil replied:
Why does it fail?

mjlbach replied:
cfg.registrationSecretFile is set to null by default, so the string substitution fails

mjlbach replied:
I just pushed an option check, let me know what you think

Infinisil replied:
Ah well you'll have to use a different start script depending on if it's set to `null`

Infinisil replied:
Note the `mkIf (cfg.registrationSecretFile != null)` in my example

Infinisil replied:
Something like
```nix
{
  script = ''
    ${if (cfg.registrationSecretFile == null) ''
      configFile=$(mktemp)
      ...
    '' else
      configFile=${configurationYaml}
    ''}
    dendrite-monolith-server ... 
  '';
}
```

Infinisil replied:
Hmm.. I'm afraid that the `sed` invocation could have escaping problems for special characters in the secret

mjlbach replied:
I was worried about that too, what do you think though of the optional check? I think it works? 

Re: sed, I'll look into seeing if there's a better option. Maybe jq? 

```
echo '{"hello": "world"}' | jq --arg foo bar '. + {foo: $foo}'
# {
#   "hello": "world",
#   "foo": "bar"
# }
```

Wulfsta replied:
Regarding `sed`, this is a bit obtuse, but maybe something like a `sed` to change `NIXOS_REGISTRATION_SECRET` to `@nixos_registration_secret@` and then a 
```
(substituteAll {
  src = "${configurationYaml}";
  nixos_registration_secret = "${registrationSecret}";
})
```
would do it? I only looked at this for about two minutes so I expect this to be a poor solution, but thought you might benefit from being pointed at `substituteAll` in this case.

mjlbach replied:
@Infinisil I think the only issue I can find is with \, but generally those aren't used in passwords anyways. @Wulfsta I don't think that solves the current issue.

Wulfsta replied:
I guess the only way to use that function would be to write up a `start-dendrite.nix` that overwrites the `shellHook` of `mkShell` and then:
```
...
settings.client_api.registration_shared_secret = "@nixos_registration_secret@";
...
script = ''
  nix-shell /path/to/start-dendrite.nix
''
...
```
Not sure that would be the best solution for this, but it was interesting to think about.

szlend replied:
How about using environmental variables and `envsubst`? That should handle most escaping issues, apart from maybe YAML's literals, which you could handle with `jq` since YAML is a superset of JSON.

I haven't tested this, but maybe something like this?

```
settings.client_api.registration_shared_secret = "$REGISTRATION_SHARED_SECRET";
```
```
REGISTRATION_SHARED_SECRET="$(cat "$secret_path" | jq -aR . )"
envsubst -i "$input"  -o "$output"
```

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,171 @@
+    systemd.services.matrix-dendrite-key-generator =
+      lib.mkIf (cfg.generatePrivateKey || cfg.generateTls ) {
+        description = "Dendrite Matrix homeserver - Key Generator";
+        wantedBy = [ "matrix-dendrite.service" ];
+        requiredBy = [ "matrix-dendrite.service" ];
```
Nit: `requiredBy` implies `wantedBy`, so you can remove the `wantedBy`

mjlbach replied:
Good to know :) Fixed

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,171 @@
+            "${pkgs.matrix-dendrite}/bin/generate-keys"
+          ] ++ lib.optionals cfg.generatePrivateKey [
+            "--private-key ${cfg.settings.global.private_key}"
+          ] ++ lib.optionals (cfg.generateTls && cfg.tlsCert && cfg.tlsKey) [
```
Assuming https://github.com/NixOS/nixpkgs/pull/109561#discussion_r563351017 is implemented, you can just use `lib.optionals cfg.generateTls` here

mjlbach replied:
Done

Wulfsta commented:
Since this is a server application, can you add systemd sandboxing to it? This might be better left as an upgrade to be made later, but it should be considered. See #93293 for similar sandboxing.

mjlbach commented:
@Wulfsta I agree with you, but I think it's probably best in it's own PR. I would have to read up on the sandboxing and would like to get this in soon-ish.

This is currently only blocked (I believe) by the sed substitution issue which I still haven't thought about/resolved.

leo60228 commented:
How do I set this up? This is my configuration:
```nix
  services.matrix-dendrite = {
    enable = true;
    httpPort = 8008;
    settings.global.server_name = "60228.dev";
  };
```
However, it just hangs:
```
Feb 07 15:37:12 leoservices systemd[1]: Starting Dendrite Matrix homeserver...
Feb 07 15:37:12 leoservices matrix-dendrite-start[8042]: time="2021-02-07T20:37:12.096054550Z" level=info msg="Dendrite version 0.3.5" func="NewBaseDendrite\n\t" >
Feb 07 15:37:12 leoservices matrix-dendrite-start[8042]: time="2021-02-07T20:37:12.112974947Z" level=info msg="Starting external Monolith listener on :8008" func=>
Feb 07 15:38:42 leoservices systemd[1]: matrix-dendrite.service: start operation timed out. Terminating.
Feb 07 15:38:42 leoservices systemd[1]: matrix-dendrite.service: Failed with result 'timeout'.
Feb 07 15:38:42 leoservices systemd[1]: Failed to start Dendrite Matrix homeserver.
Feb 07 15:38:42 leoservices systemd[1]: matrix-dendrite.service: Scheduled restart job, restart counter is at 7.
Feb 07 15:38:42 leoservices systemd[1]: Stopped Dendrite Matrix homeserver.
Feb 07 15:38:42 leoservices systemd[1]: Starting Dendrite Matrix homeserver...
Feb 07 15:38:42 leoservices matrix-dendrite-start[8141]: time="2021-02-07T20:38:42.602952054Z" level=info msg="Dendrite version 0.3.5" func="NewBaseDendrite\n\t" >
Feb 07 15:38:42 leoservices matrix-dendrite-start[8141]: time="2021-02-07T20:38:42.621508853Z" level=info msg="Starting external Monolith listener on :8008" func=>
Feb 07 15:40:12 leoservices systemd[1]: matrix-dendrite.service: start operation timed out. Terminating.
Feb 07 15:40:12 leoservices systemd[1]: matrix-dendrite.service: Failed with result 'timeout'.
Feb 07 15:40:12 leoservices systemd[1]: Failed to start Dendrite Matrix homeserver.
Feb 07 15:40:13 leoservices systemd[1]: matrix-dendrite.service: Scheduled restart job, restart counter is at 8.
Feb 07 15:40:13 leoservices systemd[1]: Stopped Dendrite Matrix homeserver.
Feb 07 15:40:13 leoservices systemd[1]: Starting Dendrite Matrix homeserver...
Feb 07 15:40:13 leoservices matrix-dendrite-start[8247]: time="2021-02-07T20:40:13.096367083Z" level=info msg="Dendrite version 0.3.5" func="NewBaseDendrite\n\t" >
Feb 07 15:40:13 leoservices matrix-dendrite-start[8247]: time="2021-02-07T20:40:13.118697969Z" level=info msg="Starting external Monolith listener on :8008" func=>
```
All requests made during this period result in a 404. I suppose this might be because I'm using this package and module on top of an older nixpkgs, but this seems like a pretty weird symptom if that's the root cause.

mjlbach commented:
> How do I set this up? This is my configuration:
> 
> ```nix
>   services.matrix-dendrite = {
>     enable = true;
>     httpPort = 8008;
>     settings.global.server_name = "60228.dev";
>   };
> ```
> 
> However, it just hangs:
> 
> ```
> Feb 07 15:37:12 leoservices systemd[1]: Starting Dendrite Matrix homeserver...
> Feb 07 15:37:12 leoservices matrix-dendrite-start[8042]: time="2021-02-07T20:37:12.096054550Z" level=info msg="Dendrite version 0.3.5" func="NewBaseDendrite\n\t" >
> Feb 07 15:37:12 leoservices matrix-dendrite-start[8042]: time="2021-02-07T20:37:12.112974947Z" level=info msg="Starting external Monolith listener on :8008" func=>
> Feb 07 15:38:42 leoservices systemd[1]: matrix-dendrite.service: start operation timed out. Terminating.
> Feb 07 15:38:42 leoservices systemd[1]: matrix-dendrite.service: Failed with result 'timeout'.
> Feb 07 15:38:42 leoservices systemd[1]: Failed to start Dendrite Matrix homeserver.
> Feb 07 15:38:42 leoservices systemd[1]: matrix-dendrite.service: Scheduled restart job, restart counter is at 7.
> Feb 07 15:38:42 leoservices systemd[1]: Stopped Dendrite Matrix homeserver.
> Feb 07 15:38:42 leoservices systemd[1]: Starting Dendrite Matrix homeserver...
> Feb 07 15:38:42 leoservices matrix-dendrite-start[8141]: time="2021-02-07T20:38:42.602952054Z" level=info msg="Dendrite version 0.3.5" func="NewBaseDendrite\n\t" >
> Feb 07 15:38:42 leoservices matrix-dendrite-start[8141]: time="2021-02-07T20:38:42.621508853Z" level=info msg="Starting external Monolith listener on :8008" func=>
> Feb 07 15:40:12 leoservices systemd[1]: matrix-dendrite.service: start operation timed out. Terminating.
> Feb 07 15:40:12 leoservices systemd[1]: matrix-dendrite.service: Failed with result 'timeout'.
> Feb 07 15:40:12 leoservices systemd[1]: Failed to start Dendrite Matrix homeserver.
> Feb 07 15:40:13 leoservices systemd[1]: matrix-dendrite.service: Scheduled restart job, restart counter is at 8.
> Feb 07 15:40:13 leoservices systemd[1]: Stopped Dendrite Matrix homeserver.
> Feb 07 15:40:13 leoservices systemd[1]: Starting Dendrite Matrix homeserver...
> Feb 07 15:40:13 leoservices matrix-dendrite-start[8247]: time="2021-02-07T20:40:13.096367083Z" level=info msg="Dendrite version 0.3.5" func="NewBaseDendrite\n\t" >
> Feb 07 15:40:13 leoservices matrix-dendrite-start[8247]: time="2021-02-07T20:40:13.118697969Z" level=info msg="Starting external Monolith listener on :8008" func=>
> ```
> 
> All requests made during this period result in a 404. I suppose this might be because I'm using this package and module on top of an older nixpkgs, but this seems like a pretty weird symptom if that's the root cause.

That's not hanging, that's the correct start message. You probably also want to set 

`client_api.registration_disabled = false`

leo60228 commented:
I don't think "start operation timed out" is correct, is it? Does the server not have full functionality until a user is created?

mjlbach commented:
It times out a couple times, but the last two lines indicate the server has started.

```
Feb 07 15:40:13 leoservices matrix-dendrite-start[8247]: time="2021-02-07T20:40:13.096367083Z" level=info msg="Dendrite version 0.3.5" func="NewBaseDendrite\n\t" >
Feb 07 15:40:13 leoservices matrix-dendrite-start[8247]: time="2021-02-07T20:40:13.118697969Z" level=info msg="Starting external Monolith listener on :8008" func=>
```

mjlbach commented:
These are the exact settings I use to test using nixos-shell:
```
{ pkgs, cfg, ... }: {
  boot.kernelPackages = pkgs.linuxPackages_latest;
  nixos-shell.mounts = {
    mountHome = false;
    mountNixProfile = false;
    cache = "none"; # default is "loose"
  };
  virtualisation.qemu.networkingOptions = [
    # We need to re-define our usermode network driver
    # since we are overriding the default value.
    "-net nic,netdev=user.0,model=virtio"
    # Than we can use qemu's hostfwd option to forward ports.
    "-netdev user,id=user.0,hostfwd=tcp::8008-:8008"
  ];
  services.matrix-dendrite.enable=true;
  services.matrix-dendrite.settings.global.server_name = "test-nixos-server";
  services.matrix-dendrite.settings.client_api.registration_disabled = false;
}
```
You then register by pointing element's homeserver at `http://localhost:8008` 
![Screenshot_20210207_125229](https://user-images.githubusercontent.com/13316262/107159247-aeb8bd00-6943-11eb-98cd-7f13a1840450.png)

leo60228 commented:
Okay, Element works. However, even with Element working, `matrix-dendrite.service` is restarting every 30 seconds due to a timeout. The example systemd service at https://github.com/matrix-org/dendrite/blob/master/docs/systemd/monolith-example.service is using `Type=simple`, is `Type=notify` really correct?

mjlbach commented:
I copied that from synpase's systemd service, I'm not quite sure of the difference tbh. From the docs:

```
notify: Behavior of notify is similar to exec; however, it is expected that the service sends a notification message via sd_notify(3) or an equivalent call when it has finished starting up. systemd will proceed with starting follow-up units after this notification message has been sent. If this option is used, NotifyAccess= (see below) should be set to open access to the notification socket provided by systemd. If NotifyAccess= is missing or set to none, it will be forcibly set to main. Note that currently Type=notify will not work if used in combination with PrivateNetwork=yes.
```

Maybe sd_notify isn't occuring, which is why the restarts is happening. If you change it to simple does everything still work for you?

leo60228 commented:
Yeah, just changing it to `simple` works fine.

mjlbach commented:
I didn't see the behavior you are seeing with the multiple restarts, but I changed the service type to simple anyways since that's what upstream ships with.

Wulfsta commented:
> @Wulfsta I agree with you, but I think it's probably best in it's own PR. I would have to read up on the sandboxing and would like to get this in soon-ish.
> 
> This is currently only blocked (I believe) by the sed substitution issue which I still haven't thought about/resolved.

Once this is merged, we should create a new issue to track adding sandboxing. Thoughts?

mjlbach commented:
> Once this is merged, we should create a new issue to track adding sandboxing. Thoughts?

Sure, I'll make the issue and likely will add sandboxing if no one beats me to it.

szlend commented on nixos/modules/services/misc/matrix-synapse.nix:
```diff
@@ -704,7 +704,7 @@ in {
       environment.PYTHONPATH = makeSearchPathOutput "lib" cfg.package.python.sitePackages [ pluginsEnv ];
       serviceConfig = {
-        Type = "notify";
+        Type = "simple";
```
I don't think you meant to edit this :)

mjlbach replied:
Whoops, thanks

mjlbach replied:
Should be fixed

szlend commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,196 @@
+            default = true;
+            description = ''
+              Disables federation. Dendrite will not be able to make
+              any outbound HTTP requeststo other servers and the federation
```
Typo in `requeststo`.

mjlbach replied:
Good catch :)

mjlbach commented:
@szlend Thanks, I just pushed and tested, it seems like it works? Does the implementation look ok to you?

mjlbach commented:
Also added some documentation for the environment secrets taken from the PR for murmur. Otherwise, please let me know what else I can do to help get this merged :)

szlend commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,208 @@
+      };
+      script = lib.concatStringsSep "\n" (
+        lib.optionals (cfg.registrationSecretFile != null) [
+          "eval export $(cat ${cfg.registrationSecretFile} | ${pkgs.jq}/bin/jq -aR . )"
```
In this case `jq` is not really helping because it was intended for escaping a single string literal, not multiple `KEY=VALUE` pairs. Anyway, I think it wasn't a big deal in the first place as it's only a problem if the secret starts with `"` or `'` which is extremely rare and can be manually escaped by the user. I only mentioned it because it looked like the review was stuck around escaping and this ensured that everything is escaped properly for a single string literal.

If you wanna go with the generic secrets file approach, then I would just remove `jq` to make things simpler as `envsubst` already solves most of the problems with `sed`.

szlend replied:
One more thing. If you're gonna use a generic env file approach, you could just use  `serviceConfig.EnvironmentFile` so you don't have to eval/export.

mjlbach replied:
That's exactly what I ended up doing :) Do you want to take a look at the latest changes? I believe this now (mostly) matches the mutter implementation.

szlend reviewed (commented):
@mjlbach I left some more comments for the recent configuration changes. Otherwise this is looking great!

rnhmjoj commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,208 @@
+        if cfg.generateTls is not set true.
+      '';
+    };
+    registrationSecretFile = lib.mkOption {
+      type = lib.types.nullOr lib.types.path;
+      example = "/var/lib/matrix-dendrite/registration_secret";
+      default = null;
+      description = ''
+          Environment file as defined in <citerefentry>
+          <refentrytitle>systemd.exec</refentrytitle><manvolnum>5</manvolnum>
+          </citerefentry>.
+          Secrets may be passed to the service without adding them to the world-readable
+          Nix store, by specifying placeholder variables as the option value in Nix and
+          setting these variables accordingly in the environment file, currently only used
+          for the registration secret. Used to allow
+          secure registration when client_api.registration_disabled is
+          true.
+          <programlisting>
+            # content of the environment file
+            REGISTRATION_SHARED_SECRET=verysecretpassword
+          </programlisting>
+          Note that this file needs to be available on the host on which
+          <literal>dendrite</literal> is running.
+      '';
+    };
```
I like this approach too, I have used in the [searx](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/services/networking/searx.nix#L66) and it's working great. Having an `environmentFile` that's included in the systemd unit is also useful for other purposes besides settings secrets.

mjlbach replied:
Addressed as of the latest commit, it now matches mutter

mjlbach commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,208 @@
+          ''
+            configFile=$(mktemp)
+            ${pkgs.envsubst}/bin/envsubst -i ${configurationYaml} -o "$configFile"
+            ${dendriteServerCmd}
```
Do I still need to do this with the changes I made?

szlend commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,207 @@
+
+          <programlisting>
+            # snippet of dendrite-related config
+            services.matrix-dendrite.settings.client_api.registration_shared_secret = $REGISTRATION_SHARED_SECRET
```
Use quotes and semicolon `"$REGISTRATION_SHARED_SECRET";` so it's a valid Nix expression.

szlend reviewed (commented):
@mjlbach From my side this looks great I only noticed a tiny documentation error. Also personally I would have used `RuntimeDirectory/RuntimeDirectoryMode` for the config file so that way it gets cleaned up after the service stops. Though it's probably not a big deal if `substenv` can overwrite the file. I'll try and test it out on my own machine in the evening.

Either way, I don't have merging rights so maybe @Infinisil could take another look now that the secret configuration problem has been resolved.

mjlbach commented:
@szlend I changed it to use runtimedirectory/runtimedirectory mode

mjlbach commented:
@szlend Did you have time to test yet?

szlend commented:
@mjlbach I just did and it works great! I only ran into one issue:

```
The option `services.matrix-dendrite.settings.global.server_name' is used but not defined.
```

This option should probably be required, right?

Anyway, hopefully someone from the NixOS team can check and get this merged if everything else is fine. @mjlbach, maybe re-request a review in the reviewers list.

mjlbach commented:
@szlend I believe that is the correct error message? I based it off of what @Infinisil suggested here https://github.com/NixOS/nixpkgs/pull/109561#discussion_r563325139

szlend commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,209 @@
+        requiredBy = [ "matrix-dendrite.service" ];
+        before = [ "matrix-dendrite.service" ];
+        path = with pkgs; [ matrix-dendrite ];
+        serviceConfig = {
+          Type = "oneshot";
+          DynamicUser = true;
+          StateDirectory = "matrix-dendrite";
+          WorkingDirectory = workingDir;
+          RemainAfterExit = true;
+          ExecStart = lib.strings.concatStringsSep " " ([
+            "${pkgs.matrix-dendrite}/bin/generate-keys"
+          ] ++ lib.optionals cfg.generatePrivateKey [
+            "--private-key ${cfg.settings.global.private_key}"
+          ] ++ lib.optionals cfg.generateTls [
+            "--tls-cert ${cfg.tlsCert}"
+            "--tls-key ${cfg.tlsKey}"
+          ]);
```
Won't this overwrite the key after a reboot? Maybe instead we should only run this if the key does not exist

```
unitConfig.ConditionPathExists = "!${cfg.settings.global.private_key}";
```

In that case I would also remove `RemainAfterExit = true` so this check is performed each time dendrite starts.

What do you think?

mjlbach replied:
Not sure what to do about this :/

szlend replied:
Maybe turn it into a shell script that generates each of those files separately when missing. Or you could just simplify the whole thing and remove TLS generation which I think was discussed previously in the PR. Self-signed certs are not very useful anyway and most people will use LetsEncrypt. In that case the `ConditionPathExists` trick would work.

kragniz replied:
I think anyone deploying this is going to be bringing their own certs anyway? I'd be keen to simplify

mjlbach replied:
Is it worth having the one-shot service for the private-key at that point?

szlend commented:
Not really. I think a preStart script would work just fine.

> On 18 Feb 2021, at 16:52, Michael Lingelbach <notifications@github.com> wrote:
> 
> ﻿
> @mjlbach commented on this pull request.
> 
> In nixos/modules/services/misc/matrix-dendrite.nix:
> 
> > +        serviceConfig = {
> +          Type = "oneshot";
> +          DynamicUser = true;
> +          StateDirectory = "matrix-dendrite";
> +          WorkingDirectory = workingDir;
> +          RemainAfterExit = true;
> +          ExecStart = lib.strings.concatStringsSep " " ([
> +            "${pkgs.matrix-dendrite}/bin/generate-keys"
> +          ] ++ lib.optionals cfg.generatePrivateKey [
> +            "--private-key ${cfg.settings.global.private_key}"
> +          ] ++ lib.optionals cfg.generateTls [
> +            "--tls-cert ${cfg.tlsCert}"
> +            "--tls-key ${cfg.tlsKey}"
> +          ]);
> Is it worth having the one-shot service for the private-key at that point?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

expipiplus1 commented:
Aside from the TLS cert issue (which imo should be removed for simplicity) is there anything stopping this from being merged?

mjlbach commented:
I don't think so, I need to rebase and bump to 0.3.11, I can do that and resolve the TLS issue later this week if there are no more comments.

ookhoi commented:
> I don't think so, I need to rebase and bump to 0.3.11, I can do that and resolve the TLS/private key issue (remove it) later this week if there are no more comments.

As a positive comment, at my very small familyserver I've replaced `matrix-synapse` with `matrix-dendrite` from your branch running at `0.3.11`.

Thanks a lot for your work!

```
diff --git a/pkgs/servers/matrix-dendrite/default.nix b/pkgs/servers/matrix-dendrite/default.nix
index a1e325b8e1c..708a1e6f65e 100644
--- a/pkgs/servers/matrix-dendrite/default.nix
+++ b/pkgs/servers/matrix-dendrite/default.nix
@@ -2,16 +2,16 @@

 buildGoModule rec {
   pname = "matrix-dendrite";
-  version = "0.3.9";
+  version = "0.3.11";

   src = fetchFromGitHub {
     owner = "matrix-org";
     repo = "dendrite";
     rev = "v${version}";
-    sha256 = "cJDUuQyX4JdAA/BBbnjU4h6NGs0HDTOwClTYzT/6TOQ=";
+    sha256 = "15xqd4yhsnnpz5n90fbny9i8lp7ki2z3fbpbd8cvsp49347rm483";
   };

-  vendorSha256 = "Fk9DxEAoyAK9hOchFjF3vn0SLEXUGhWUO4GmpTbOjPM=";
+  vendorSha256 = "1l1wydvi0yalas79cvhrqg563cvs57hg9rv6qnkw879r6smb2x1n";

   meta = with lib; {
     homepage = "https://matrix.org";
```

mjlbach commented:
That's wonderful :) I got behind on work last week, but I can do the bump/cleanups this weekend

mjlbach commented:
I think I have addressed everyone's concerns, users are now required to generate and specify their own matrix keys and TLS keys (no magic service anymore). I didn't add additional sandboxing but will file an issue after this is merged.

ookhoi commented:
About to test your latest changes.

One question: why does `disable_federation` default to `true` ?

The suggested default seems to be `false`
https://github.com/matrix-org/dendrite/blob/464b908bd0c13854b3f6b9a17467f39e0608dc08/dendrite-config.yaml#L65

mjlbach commented:
Ah, I'm not sure that particularly necessary to have as a default option, so I just went ahead and removed the whole thing (it can still be set by the user ofc)

jacereda commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,162 @@
+      default = null;
+      description = ''
+        The path to the TLS certificate. Will use pre-existing cert
+        if cfg.generateTls is not set true.
```
Is this documentation obsolete now?

mjlbach replied:
Done

jacereda commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,162 @@
+      default = null;
+      description = ''
+        The path to the TLS key. Will use pre-existing key
+        if cfg.generateTls is not set true.
```
Same here.

mjlbach replied:
Done

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,160 @@
+          "--config /run/matrix-dendrite/dendrite.yaml"
+        ] ++ lib.optionals (cfg.httpPort != null) [
+          "--http-bind-address :${builtins.toString cfg.httpPort}"
+        ] ++ lib.optionals (cfg.httpsPort != null) [
+          "--https-bind-address :${builtins.toString cfg.httpsPort}"
+        ] ++ lib.optionals (cfg.tlsCert != null && cfg.tlsKey != null) [
+          "--tls-cert ${cfg.tlsCert}"
+          "--tls-key ${cfg.tlsKey}"
```
How are these options related? Can HTTPS be enabled without giving a TLS cert/key? Maybe an assertion like
```nix
{
  assertions = [{
    assertion = httpsPort != null -> (tlsCert != null && tlsKey != null);
    message = "...";
  }];
}
```

is appropriate. And then you could also do
```suggestion
        ] ++ lib.optionals (cfg.httpsPort != null) [
          "--https-bind-address :${builtins.toString cfg.httpsPort}"
          "--tls-cert ${cfg.tlsCert}"
          "--tls-key ${cfg.tlsKey}"
```

mjlbach replied:
Fixed, you need to provide tls-cert/tls-key if you bind https. I added additional instructions for generating these keys.

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,160 @@
+            example = "${workingDir}/matrix_key.pem";
+            description = ''
+              The path to the signing private key file, used to sign
+              requests and events.
```
What does a value of `null` mean? Should probably be in the description

mjlbach replied:
It's a remnant of when I used to have a service to generate this key for you. Removed.

mjlbach commented:
@Infinisil I think I addressed your review.

Infinisil commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,181 @@
+    assertions = [{
+      assertion = cfg.httpsPort != null -> (cfg.tlsCert != null && cfg.tlsKey != null);
+      message = ''
+        If Dendrite is configured to use https without a reverse proxy, tlsCert and tlsKey must be provided.
```
```suggestion
        If Dendrite is configured to use https, tlsCert and tlsKey must be provided.
```

Or does using a reverse proxy really not require a TLS key + cert?

mjlbach replied:
I changed it, I'm not sure how it works to be honest, I haven't tried using a reverse proxy for dendrite, but my understanding was the TLS key of the proxy was used instead.

Infinisil replied:
Yeah I also believe so.

infinisil reviewed (approved):
Looking good!

mjlbach commented:
Nice! Feel free to merge whenever :) (I can't)

dotlambda commented on pkgs/servers/matrix-dendrite/default.nix:
```diff
@@ -0,0 +1,23 @@
+  vendorSha256 = "1l1wydvi0yalas79cvhrqg563cvs57hg9rv6qnkw879r6smb2x1n";
+
+  meta = with lib; {
+    homepage = "https://matrix.org";
```
This doesn't seem appropriate.

mjlbach replied:
If so, then synapse should also be changed https://github.com/NixOS/nixpkgs/blob/1d64c0db8524b94c73e5a3d18e3c1c21c2ac0396/pkgs/servers/matrix-synapse/default.nix#L76-L78

dotlambda replied:
Yes, I think so

dotlambda commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,181 @@
+              -i ${configurationYaml} \
+              -o /run/matrix-dendrite/dendrite.yaml
+          '' else ''
+            ${pkgs.coreutils}/bin/cp ${configurationYaml} /run/matrix-dendrite/dendrite.yaml
```
Can we use `ln -s`?

mjlbach replied:
I copied the behavior of the branch with `${pkgs.envsubst}/bin/envsubst` which introduces a "secret" into the file. I think it's better the behavior of both branches match

dotlambda commented on nixos/modules/services/misc/matrix-dendrite.nix:
```diff
@@ -0,0 +1,181 @@
+        EnvironmentFile = lib.mkIf (cfg.environmentFile != null) cfg.environmentFile;
+        ExecStartPre =
+          if (cfg.environmentFile != null) then ''
+            ${pkgs.envsubst}/bin/envsubst \
+              -i ${configurationYaml} \
+              -o /run/matrix-dendrite/dendrite.yaml
```
Can we use chmod to make it non-writable?

mjlbach replied:
I think so, this can be done in the sandboxing PR

dotlambda commented:
Why was this even merged without tests?

Infinisil commented:
Yeah tests would be ideal, but not required imo. Also this PR is already pretty old and has seen a lot of review, the fact that @mjlbach has been updating it throughout all this time is commendable. But I think at some point the potential risk of losing a valuable contributor just because of PR's being reviewed for too long over-weighs the benefits.

mjlbach commented:
I have a WIP branch with tests, but as this is my first time writing tests for a nixos module, I am having a bit of difficulty debugging. I'm happy to open the draft PR today for feedback, and will address additional feedback to continue to improve the module. FWIW I think this is a valuable module as-is as it makes matrix homeserver hosting much more accessible (synapse is a resource hog, and is infeasible to run on a raspberry pi as many do with znc bouncers.

@Infinisil I was happy to stick through the review process, and thank you for continuing to provide feedback throughout it as well :)

mjlbach commented:
As an update, the tests are blocked by the fact that matrix python sdk which we use for tests hasn't been updated in 2 years and isn't compatible with dendrite. Matrix-nio isn't either, but I submitted an upstream patch which will fix it. https://github.com/poljar/matrix-nio/pull/254

I'm going to go ahead and open the draft PR, pending that fix being merged and the version bumped in nixpkgs

dotlambda commented:
> Yeah tests would be ideal, but not required imo.

They are not required but it seems like in this case it would have been rather easy to copy them from Synapse's.

> But I think at some point the potential risk of losing a valuable contributor just because of PR's being reviewed for too long over-weighs the benefits.

 100% agreed

> FWIW I think this is a valuable module as-is as it makes matrix homeserver hosting much more accessible

For sure! If only dendrite supported push notifications (not for me, but for those who use crippled phones).

mjlbach commented:
> They are not required but it seems like in this case it would have been rather easy to copy them from Synapse's.

Yes, sadly this was not the case as the python SDK used by those tests is fairly [abandoned](https://github.com/matrix-org/matrix-python-sdk). Matrix-nio, which I used for the new test, supports dendrite (as of tonight) and probably should be preferred for testing all matrix-servers going forward.
